src/ripple/rpc/handlers/AccountInfo.cpp

```diff
@@ -183,4 +186,94 @@ Json::Value doAccountInfo (RPC::Context& context)
     return result;
 }
 
+std::pair<rpc::v1::GetAccountInfoResponse, grpc::Status> doAccountInfoGrpc(
+        RPC::ContextGeneric<rpc::v1::GetAccountInfoRequest>& context)
+{
+    //Return values
```

Comment:
I totally see where you're coming from, and I originally wanted to do that. However, I think that would require modifying a lot of existing code, as a lot of the helper functions are tightly coupled with json; for example, when I attempted to do this, there were helpers that called helpers that called helpers, that all took json as arguments and returned json.

I'm hesitant to modify so many existing code paths, because I think it would greatly increase the chance of breaking existing functionality. Though I do agree that keeping the API's in sync would be a pain, and having common handlers would be optimal. Do we think refactoring a bunch of these handlers and helper functions is a good idea, even though that would modify a lot of existing code paths? There are already quite a few changes in this PR as is, but I would be happy to refactor if people think it is a good idea. Would like to hear some more opinions.

Reply:
This is a great comment @seelabs ! We struggled with this during the design as well. Here are my thoughts: 
1. The ideal solution if we were to start new is to obviously decouple the output generation (json and proto) from the actual command handling, and then we can use a common handler to handle all interfaces. But in this case, I felt that the amount of risk in modifying existing handlers are too high for the initial gRPC deployment. If we modify the existing handlers, like Cj said there are a lot of helper functions that need to be updated as well and we essentially increase the surface of the code that we touch - and obviously there will be unit test impact as well. 
2. I think we will very quickly upgrade the gRPC handlers (v2 per our discussion) and at that point we will have different handlers since I do not think we will upgrade the existing JSON handlers . Note this does not mean we will not have V2 JSON support - gRPC allows the generation of JSON from proto so that we can also provide JSON V2 its just that will not be the same handlers. The V2 JSON and V2 gRPC would be handled by common handlers and the process of translating proto <-> JSON would be handled by protobuf. (essentially we get JSON for free). Its at that point we can consider the existing JSON as V1 handlers and the new improved (also peer reviewed with spec) V2 handlers can support both JSON and gRPC with common handlers.
3. I also think this deployment is somewhat “undocumented” upgrades in the sense that this can change in the future. So it's better not to “contaminate” the existing handlers.

@nbougalis  & Peng any thoughts?

Reply:
I agree with 1 as we discussed that during the design. 

About 2, just to be clear, we do not have to quickly upgrade gRPC handlers to V2, unless we need to modify the current gRPC API in a breaking way. We will add new functions or adding new fields of existing functions in the current version of gRPC interface, but those are not breaking changes, so we don't have to advance the version number. I imagined we will stay in V1 for awhile till we finishes all the missing functions. A minor point about Json and gRPC transcoding, I think I saw some of the current json names' formats are not consistent (snake vs camel, etc). We may want to change it in V2, this need to discussed and agreed.

BTW, I think this PR is different from the other issues we worked on so far since it introduces a sizable new component in rippled code. We should have emphasized the design so its discussed in design review.

Reply:
@cjcobb23 @pwang200 @carlhua

I understand the motivation for duplicating the handlers. I'm OK with doing this for a prototype implementation, but not in a production implementation. The question is, is this a prototype, or this is production? Let me make the argument that this is a prototype, and if we can agree to that, then I'm OK duplicating the handlers. If we can't agree, then we have to think about this some more.

I talked to @carlhua about some concerns I had about the API proposed here. I am concerned that the API is not ready to be baked into rippled as an official API. We already have a JSON API, and it appears this gRPC API will use the JSON API as a reference. Even trying to match the existing JSON API may be hard to get right the first time. There were already some issues in the protobufs API (that were caught in review), but given this is being developed incrementally, I'm not comfortable that we would only make backward compatible changes to this proposed partial API.

In addition, we don't often get a chance to change APIs. Both Elliot and Rome have dealt with our current API and would be good people to help improve a new API. 

After talking with Carl, We decided this API would only be used by Keefer, and we would not commit to backward compatibility. After this version was iterated on and all the APIs were available, a we would release a new official API (after the prototype API had some internal use and was reviewed by Rome, Elliot, and others). Once the official version was released, the prototype version would be removed from rippled entirely (with perhaps a short period - one version - supporting both prototype and release gRPC APIs to allow for a transition).

In order for this to work, I think we need at least the following:

1) Buy-in from @keefertaylor that this API is experimental, and anything that's built on it would need to be changed when the production API was released.
2) Build the code in such a way it is easy to remove from rippled.
3) By default rippled servers would not turn on the experimental gRPC API. The new API would have to be opt-in and in such a way that it's clear this is a prototype that may change.
4) I don't want to commit to supporting this prototype API indefinitely if the production API is difficult to agree on.
5) We have to agree that it's realistic to remove this API.
   

I'm not sure how realistic that we'll actually be able to remove an API once it's in rippled. My preference is still to build this API outside of rippled and potentially integrate it later (build a gRPC server that talks to rippled through the rippled's JSON API). However, I understand we can build prototypes faster by integrating it into rippled. If we're integrating it in, let's at least do everything we can to signal that it's a prototype.

If we can't agree that this is a prototype, then, well, we'll have to talk some more and I'll try to articulate why I object to shipping this as a production implementation.

Reply:
@seelabs @carlhua @pwang200 I'm OK with this API being a prototype. I think trying to create a production ready gRPC API in one pull request is not the best idea; it's hard to get everything right, and we don't want to commit to every single nuance of the API up front. I think it's better to create a minimal working version (as this PR is), release it as a prototype, and allow time for internal use and additional feedback before creating a production ready gRPC API.

About each of your points:
1. defer to @keefertaylor . For what it's worth, I don't think the production API would be a dramatic overhaul of what is currently here (after implementing the requested changes), but would mostly consist of additions (support more commands and fully support the `tx` command), and small changes.
2. Given the way the code is currently written, removing this from rippled would be easy. Very few existing code paths were modified, the gRPC handlers are separate from JSON, and the entire gRPC server code is contained in two files. 
3. I agree that the gRPC API should be opt-in, and disabled by default.
4. I think releasing the gRPC API as a prototype absolves us from supporting the gRPC API indefinitely; prototypes can be removed/changed at any time.
5. I agree that it's realistic to remove this API. Since this is a prototype, we can decide later if we want to productionize or remove it.

> I'm not sure how realistic that we'll actually be able to remove an API once it's in rippled.

@seelabs What issues do you see with removing the API? It seems to me that removal would really only impact @keefertaylor, and, if he still needs this functionality, we can just setup a gRPC server that talks JSON to rippled, for the commands that Keefer needs for the SDKs. Keefer actually already wrote something like that.

Reply:
Agree with everything said. 

I think that this is a good first pass at our gRPC API. I think that we should do everything in our power to do a proper API review now, and ideally there are no breaking changes before production, but I don't object to leaving that as a possible option. 

The Xpring SDK work I'm doing encapsulates the network calls completely so we can accommodate breaking changes if needed.

Reply:
@seelabs, @cjcobb23, @carlhua, @pwang200, please excuse me sticking my nose in.  If we agree that what is being built is a prototype, then perhaps some simplifying assumptions can be made.

1. The branch with these changes will work just fine on a network with production rippled servers.  There is no need to merge the prototype into rippled's develop branch in order to test and develop your prototype.

2. If you are not working to merge the prototype into rippled's develop branch, then the code reviews and unit test requirements can be significantly reduced or removed.  This will allow the prototype to progress faster.

3. When the initial prototype is done, consider building the next step as a process outside of rippled as @seelabs suggests.  This will give you much more freedom to follow your own coding standards and expectations.  It also introduces significantly less risk and maintenance overhead into the rippled codebase.

Just for your consideration.

Reply:
As I understood it, stuff in this PR is needed for xSpring's SDK which already has a (proprietary) middleware written as a translation layer between gRPC + JSON-RPC and this is the attempt to get gRPC directly into `rippled` instead of releasing or refining that middleware.

By the way: `rippled` APIs are currently not versioned and also not marked as "experimental" or "unstable", see the discussion in https://github.com/ripple/rippled/pull/3155 where it is argued that everything that ships is production ready and production quality.

Reply:
@scottschurr @seelabs Here are my thought: This PR can be merged in as experimental/prototype and we can anticipate changes in the future. We can put comments to say so. We can also expect the only user of such interface is the Xpring SDK. In addition, I am in agreement to make this an opt-in feature in such a way that only when its explicitly enabled in the config - then the feature will be turned on. This way by default, there is virtually no impact to other people. However, if this is not merged into the master and only lives on a branch - then it will not work because then the general public will not get access to gRPC via Xpring SDK inside rippled even if they wanted to. With that said, I think its imperative for us to support Xpring SDK and its development. Since I see that there is little impact with such proposal I suggest that we go with this route. @keefertaylor  do you have any comments?
If we still disagree, then IMO this becomes a blocking issue that goes beyond the current code review's scope.  please do let me know so I can help sorting this out.

I would also like to use this opportunity to voice some of my thoughts that are tangential: For new features such as gRPC, at least during a code review - there shouldn't be any design related comments or comments that argue "whether the feature should have been developed". As part of the software engineering process, IMO a code review is only to assess whether an implementation has correctly realized its associated design. Arguments for whether something should have been done should occur prior to the code review. In addition, the design of the implementation for new feature should also be discussed and reviewed prior to code reviews. In this case, the design document was presented internally ahead of the code review. (note, the usage of individual handlers vs common handlers are explicitly stated in the design as well) I understand that this is @cjcobb23 first time developing new features for rippled and the processes/expectations aren't perfectly aligned (I should have made it more clear on the expectation of design document) but I would certainly love to see that we consider this particular thought of mine. I am happy to chat if anyone has any concerns.

Reply:
Agree. Xpring SDK (And proprietary middleware) is the prototype. The next phase is to move away from our proprietary software so that folks don’t need to rely on/trust Ripple Labs / Xpring to use our client software. We’re definitely past the prototype phase. 

There is value in having this API in Rippled as it let’s us continue to gather data points about performance and stability (in both the client SDK and the node). It also allows enthusiastic or adventurous community members to provide early feedback. 

I am in favor of saying this API is experimental and opt in at least for n+1 after we have parity with the JSON API. After that, I’m still fine with opt in. The experimental stays should come with a disclaimer that we may make breaking API changes. However, I would hope that our reviews are rigorous and the design thought out enough that we’ll never have to actually make a breaking change but we will have that option. 

In short, please review this like it was going to be labeled as a stable API that we’ll support indefinitely as that’s the standard we want this code to have.

Reply:
I understand @seelabs's concern about code duplication; in fact I think we all do. We should move to a single, integrated handler.

With that said, integrating **gRPC** support is likely to be a process, and this is the first step. Making extensive changes, including dramatic alterations to handlers, at this point seems like a bad idea.

I think there's value in merging this PR and further work on this in stages, as opposed to maintaining it in a branch and then doing one huge change.

So, let's augment this PR to add a configuration option that must be passed in to _explicitly enable_ the **gRPC** support at runtime, allowing those that need it or want to test it to run it. Then we can improve things as we go along, while modulating the impact on the codebase.

Reply:
@nikb should the configuration option be passed via command line or can it be set in the configuration file? I feel passing via command line is more explicit

Reply:
config file seems obvious - if there is no port entry using the grpc protocol, then you don't start any grpc server instances.  That said, I'm also in favor (at least for now) of a build/config time option so that it's possible to build a version of rippled that has no dependency on grpc whatsoever. We'd have to figure out how such a build option should translate into packaging variants.

Reply:
changed to using the config file to determine whether to start gRPC or not